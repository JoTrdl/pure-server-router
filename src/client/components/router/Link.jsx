import React from 'react'
import PropTypes from 'prop-types'

/**
 * Router Link component
 *
 * @example
 * <Link to="">
 * <Lint to="" active>
 * <Link to="" match>
 * <Link to="" exact>
 * <Link to="" exact withQueries>
 * <Link to="" exact withQueries=['sort']>
 *
 * @class Link
 * @extends {React.Component}
 * @reactProps {boolean} active - Force the link active state
 * @reactProps {string} activeClassName - The classname to use for 'active' state
 * @reactProps {boolean} match - Compare the current URL with match()
 * @reactProps {boolean} exact - Compare the current URL with equal (===)
 * @reactProps {boolean|Array} withQueries - Include queries in the test comparaison
 */
export default class Link extends React.Component {
  static contextTypes = {
    router: PropTypes.object
  }

  componentWillMount () {
    this._unlisten = this.context.router.listen(() => this.forceUpdate())
  }

  componentWillUnmount () {
    if (this._unlisten) {
      this._unlisten()
    }
  }

  render () {
    const { router } = this.context
    const { location } = router
    const {
      to, exact, match, withQueries,
      active, activeClassName, className,
      children, onClick, ...otherProps } = this.props

    const route = to || '#'
    const [path, qs] = route.split('?', 2)
    let classNames = className || ''

    let routeMatch = exact
      ? location.path === path
      : match
        ? location.path.match(path)
        : active

    // If 'withQueries' is passed, determine if it is
    // a route match based on thse queries.
    // Note: queries order is not important.
    if (withQueries) {
      const queries = (qs || '').split('&').reduce((acc, s) => {
        const [q, v] = s.split('=')
        acc[q || ''] = v || ''
        return acc
      }, {})

      const expectedQueries = Array.isArray(withQueries)
        ? withQueries
        : Object.keys(queries)

      routeMatch = expectedQueries.every(q => (
        queries[q] === `${(location.query || {})[q]}`
      ))
    }

    // If the route matches, add the 'activeClassName'.
    // Default to 'active' if nothing provided.
    if (routeMatch) {
      classNames += ` ${activeClassName || 'active'}`.trim()
    }

    // Override or set the onClick handler.
    // Will called the initial one if provided and then,
    // call push() on the router.
    otherProps.onClick = function navigate (e) {
      if (e.button !== 0 || e.ctrlKey || e.altKey || e.metaKey || e.shiftKey) {
        return
      }

      e.preventDefault()
      typeof onClick === 'function' && onClick(e)

      router.push(route)
    }

    return (
      <a href={route}
        className={classNames}
        {...otherProps}>
        {children}
      </a>
    )
  }
}
