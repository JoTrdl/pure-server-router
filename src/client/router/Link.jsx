import React from 'react'
import PropTypes from 'prop-types'

/**
 * Supports:
 *
 * <Link to="">
 * <Lint to="" active>
 * <Link to="" match>
 * <Link to="" exact>
 * <Link to="" exact withQueries>
 * <Link to="" exact withQueries=['sort']>
 * 
 * active: set the active classname
 * match: match the current location with the to
 * exact: strictly compares the location and the to
 * withQueries: compare all queries too
 * withQueries(array): compare only the specified queries
 */
export default class Link extends React.Component {
  static contextTypes = {
    router: PropTypes.object
  }

  componentWillMount () {
    this.unlisten = this.context.router.listen(() => this.forceUpdate())
  }

  componentWillUnmount () {
    if (this.unlisten) {
      this.unlisten()
    }
  }

  render () {
    const { router } = this.context
    const { location } = router
    const {
      to, exact, match, withQueries,
      active, activeClassName, className,
      children, onClick, ...otherProps } = this.props

    const route = to || '#'
    const [path, qs] = route.split('?', 2)
    let classNames = className || ''

    let routeMatch = exact
      ? location.path === path
      : match
        ? location.path.match(path)
        : active

    // If 'withQueries' is passed, determine if it is
    // a route match based on thse queries.
    // Note: queries order is not important. 
    if (withQueries) {
      const queries = (qs || '').split('&').reduce((acc, s) => {
        const [q, v] = s.split('=')
        acc[q || ''] = v || ''
        return acc
      }, {})

      const expectedQueries = Array.isArray(withQueries)
        ? withQueries
        : Object.keys(queries)

      routeMatch = expectedQueries.every(q => (
        queries[q] === `${(location.query || {})[q]}`
      ))
    }

    // If the route matches, add the 'activeClassName'.
    // Default to 'active' if nothing provided.
    if (routeMatch) {
      classNames += ` ${activeClassName || 'active'}`.trim()
    }

    // Override or set the onClick handler.
    // Will called the initial one if provided and then,
    // call push() on the router.
    otherProps.onClick = function navigate (e) {
      if (e.button !== 0 || e.ctrlKey || e.altKey || e.metaKey || e.shiftKey) {
        return
      }

      e.preventDefault()
      typeof onClick === 'function' && onClick(e)

      router.push(route)
    }

    return (
      <a href={route}
        className={classNames}
        {...otherProps}>
        {children}
      </a>
    )
  }
}
