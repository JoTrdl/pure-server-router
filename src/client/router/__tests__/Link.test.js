import React from 'react'
import PropTypes from 'prop-types'
import Link from '../Link'
import ReactTestRenderer from 'react-test-renderer'

describe('Link', () => {
  const getStubbedLink = (path, query = {}) => {
    class Wrapper extends React.Component {
      static childContextTypes = {
        router: PropTypes.object
      }

      getChildContext () {
        return {
          router: {
            location: { path, query },
            listen: () => {}
          }
        }
      }

      render () {
        return React.createElement(Link, this.props)
      }
    }

    return Wrapper
  }

  it('renders a link without active class', () => {
    const Link = getStubbedLink('/')
    const tree = ReactTestRenderer.create(
      <Link to="/test">Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link without active class even if route match', () => {
    const Link = getStubbedLink('/test')
    const tree = ReactTestRenderer.create(
      <Link to="/test">Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link with active class if route exact', () => {
    const Link = getStubbedLink('/test')
    const tree = ReactTestRenderer.create(
      <Link to="/test" exact>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link with active class if route match', () => {
    const Link = getStubbedLink('/test/sub')
    const tree = ReactTestRenderer.create(
      <Link to="/test" match>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link without active class if route match', () => {
    const Link = getStubbedLink('/test/sub')
    const tree = ReactTestRenderer.create(
      <Link to="/test/other" match>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link with active class if queries match', () => {
    const Link = getStubbedLink('/test', {a: 1, b: 2})
    const tree = ReactTestRenderer.create(
      <Link to="/test?a=1&b=2" exact withQueries>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link with active class if queries match', () => {
    const Link = getStubbedLink('/test', {a: 1, b: 2})
    const tree = ReactTestRenderer.create(
      <Link to="/test?a=1&b=3" exact withQueries={['a']}>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a link without active class if queries not match', () => {
    const Link = getStubbedLink('/test', {a: 4, b: 3})
    const tree = ReactTestRenderer.create(
      <Link to="/test?a=1&b=3" exact withQueries={['a']}>Test</Link>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
