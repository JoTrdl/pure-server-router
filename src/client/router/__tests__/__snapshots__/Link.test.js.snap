// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders a link with active class if queries match 1`] = `
<a
  className="active"
  href="/test?a=1&b=2"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link with active class if queries match 2`] = `
<a
  className="active"
  href="/test?a=1&b=3"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link with active class if route exact 1`] = `
<a
  className="active"
  href="/test"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link with active class if route match 1`] = `
<a
  className="active"
  href="/test"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link without active class 1`] = `
<a
  className=""
  href="/test"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link without active class even if route match 1`] = `
<a
  className=""
  href="/test"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link without active class if queries not match 1`] = `
<a
  className=""
  href="/test?a=1&b=3"
  onClick={[Function]}
>
  Test
</a>
`;

exports[`Link renders a link without active class if route match 1`] = `
<a
  className=""
  href="/test/other"
  onClick={[Function]}
>
  Test
</a>
`;
